### PR TITLE
boards: add the Adafruit Trinket M0.

### PR DESCRIPTION
**Pull request by nzmichaelh**
The Trinket M0 is a tiny SAMD21 based board with USB, UART, SPI, and
an on-board RGB LED.

Signed-off-by: Michael Hope <mlhx@google.com>

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]

File: boards/arm/adafruit_trinket_m0/adafruit_trinket_m0.dts
Comment (ydamigos) [inline boards/arm/adafruit_trinket_m0/adafruit_trinket_m0.dts]: DT binding for this device is missing.

Comment (ydamigos) [reply boards/arm/adafruit_trinket_m0/adafruit_trinket_m0.dts]: Ok, PR #6209 provides the binding

Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/adafruit_trinket_m0.dts]: Yip :)

**Comment by nzmichaelh** [comment]
I asked Adafruit and they said it was OK to use their photo for the board image.  Updated.

File: boards/arm/adafruit_trinket_m0/dts.fixup
Comment (galak) [inline boards/arm/adafruit_trinket_m0/dts.fixup]: Lets refactor the atmel sam0 dts.fixup into arch/arm/soc/

**Review by galak — CHANGES_REQUESTED** [review]
Looks good overall, lets refactor dts.fixup to remove duplication (see commit d76858f6ae91c325c59ff206977e60be2f384251 for an example)

File: boards/arm/adafruit_trinket_m0/dts.fixup
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/dts.fixup]: Done.  Note that the samd/ version wasn't getting picked up so I moved it to samd21/

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: ``An RGB``   (and yes, English rules for using ``a`` vs. ``an`` are tricky, based on the sound of the next word)

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: ``in the Kconfig file``

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: More descriptive if you write this as (and adjust the label name at the end of the doc): 

    The `Adafruit Trinket M0 Learn site`_ has ...

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: You use the term ``tapping`` below, I like that better since it implies a press and release:

    by quickly tapping the reset button twice.

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: maybe add:

    (See the `Serial Port`_ section above for the board's pin connections.)

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: update this label with the suggestion above (where it's referenced)

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (dbkinder) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Maybe add ``ARM`` before ``development board`` (to let folks know the architecture early on)

**Review by dbkinder — CHANGES_REQUESTED** [review]
thanks for the doc.  some suggested edits

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Done

**Review by dbkinder — APPROVED** [review]
Thanks!

**Comment by nzmichaelh** [comment]
@galak I've refactored the dts fixup to arch/arm/soc/...  I also dropped the LED strip driver as the new SPI master checks will fail until #5865 is merged

**Comment by henrikbrixandersen** [comment]
Shouldn't `dts.fixup` also include definitions for SPI (`CONFIG_SPI_SAM0_SERCOM0_BASE_ADDRESS` and friends)?

**Comment by nzmichaelh** [comment]
Hi @henrikbrixandersen my SPI in Device Tree (#5865) PR needs to land first.

I'm hoping that this PR gets approved then we can continue with the SPI related changes in the SPI PR.

**Comment by henrikbrixandersen** [comment]
Hi,

Ok, thank you. I have a port for the Adafruit Feather M0 Basic Proto ready - but waiting for this PR to land before submitting it since it too would need a refactoring af the dts.fixup.

**Comment by nzmichaelh** [comment]
Cool!  If you happen to get a Feather M0 Adalogger in the future then you might be interested in the #5528 SD card driver.

File: boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst
Comment (henrikbrixandersen) [inline boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: s/Tinket/Trinket/

Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/adafruit_trinket_m0.rst]: Good spotting, fixed and ran a spell check on the file.

**Review by galak — CHANGES_REQUESTED** [review]
Some minor cleanups / refactoring and question about change to led_ws2812 driver.

File: samples/drivers/led_ws2812/src/main.c
Comment (galak) [inline samples/drivers/led_ws2812/src/main.c]: Is this change related to the board code?  At min we need a comment in commit message about it.  Ideally it would be a separate commit.

File: arch/arm/soc/atmel_sam0/samd21/dts.fixup
Comment (galak) [inline arch/arm/soc/atmel_sam0/samd21/dts.fixup]: Can you pull the dts.fixup move/additions into their own commit in the PR.  I'd like them separated from the board code commit if possible.

File: samples/drivers/led_ws2812/src/main.c
Comment (nzmichaelh) [reply samples/drivers/led_ws2812/src/main.c]: Oops, fixed.

File: arch/arm/soc/atmel_sam0/samd21/dts.fixup
Comment (nzmichaelh) [reply arch/arm/soc/atmel_sam0/samd21/dts.fixup]: Done as #6305

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
Hmm, from shippable:

../../../../../../../include/arch/arm/cortex_m/cmsis.h:132:2: error: #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
 #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
  ^~~~~

**Comment by galak** [comment]
> Hmm, from shippable:
> 
> ../../../../../../../include/arch/arm/cortex_m/cmsis.h:132:2: error: #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
> #error "CONFIG_NUM_IRQ_PRIO_BITS and __NVIC_PRIO_BITS are not set to the same value"
> ^~~~~

Looks like we just need to update against master to get the dts.fixup

**Comment by nzmichaelh** [comment]
@galak I've rebased against master.

**Pull request merged**
